# Rede Neural Base de Crédito
#### Criando uma rede neural para resolver o problema da base de crédito.

### Pré-processamento dos dados

In [7]:
# importando a biblioteca pandas do python
import pandas as pd

# importando a base de dados para o objeto "dataframe"
dataframe = pd.read_csv('Dados de Credito.csv', encoding = 'utf-8', sep = ',')

# substituindo valores inconsistentes no campo idades pela média das idades consistentes
dataframe.loc[dataframe.age < 0, 'age'] = 40.92

# separando os atributos previsores do meta classe
previsores = dataframe.iloc[:, 1:4].values
classe = dataframe.iloc[:, 4].values

# importando a biblioteca sklearn do python
from sklearn.impute import SimpleImputer
# função "SimpleImputer" responsável por corrigir valores faltantes na base de dados

# importando a biblioteca numpy do python
import numpy as np

# criando o objeto "imputer"
imputer = SimpleImputer(missing_values = np.nan, strategy = "mean")

# fazendo o treinamento com a base de dados para correção de valores faltantes
imputer = imputer.fit(previsores[:,0:3])

# corrigindo os valores faltantes usando o objeto criado "imputer"
previsores[:, 0:3] = imputer.transform(previsores[:, 0:3])

# importando a biblioteca sklearn do python
from sklearn.preprocessing import StandardScaler
# a função "StandardScaler" tem a função de escalonas toda a base de dados para corrigir a discrepância
# dos valores

# criando o objeto "scaler"
scaler = StandardScaler()

# escalonando todos os atributos para auxiliar no cálculo de distâncias euclidianas
previsores = scaler.fit_transform(previsores)

# importando a biblioteca sklearn do python
from sklearn.model_selection import train_test_split
# a função "train_test_split" tem a importância de separar modelos de treinamento e modelos de teste
# em uma base de dados

# criando um modelo de treinamento e um modelo de teste
previsores_treinamento, previsores_teste, classe_treinamento, classe_teste = train_test_split(previsores,
                                                                                             classe,
                                                                                             test_size = 0.25,
                                                                                             random_state = 0)

### Criando uma rede neural usando a linguagem de programação Python

In [8]:
# importando a biblioteca sklearn do python
from sklearn.neural_network import MLPClassifier
# 'MLPClassifier' é usado para criar uma rede neural
# Multi Layer Perceptron

In [9]:
# criando e definindo um 'classificador' para definir os parâmetros da rede neural
# verbose mostra na tela os dados em cada época de treinamento
# max_iter = números de repetições, ou seja, quantidade de vezes que o algoritmo será treinado
# tol = ajusta os pesos com passos pequenos
# solver é a otimização utilizada, 'adam' é um melhoramente do método gradiente descente
# activation = função de ativação
# hidden_layer_sizes = quantidade de neurônios em cada camada oculta
# learning_rate = indica o método de ajuste de pesos, em 'constant' o treinamento para quando o peso não muda por duas vezes consecutivas e
# 'adaptative' muda a taxa de aprendizado quando o erro não diminui
# muda a taxa de aprendizagem
classificador = MLPClassifier(verbose = True, max_iter = 1000, tol = 0.0000010, solver = 'adam', hidden_layer_sizes = (100, 100),
                              activation = 'relu', learning_rate = 'constant', random_state = 0)

In [10]:
# fazendo o treinamento com a base de dados de treinamento
classificador.fit(previsores_treinamento, classe_treinamento)

Iteration 1, loss = 0.68988648
Iteration 2, loss = 0.48556666
Iteration 3, loss = 0.36127704
Iteration 4, loss = 0.28295103
Iteration 5, loss = 0.22659304
Iteration 6, loss = 0.18679249
Iteration 7, loss = 0.15799548
Iteration 8, loss = 0.13670874
Iteration 9, loss = 0.12083605
Iteration 10, loss = 0.10894405
Iteration 11, loss = 0.09896763
Iteration 12, loss = 0.09104808
Iteration 13, loss = 0.08438450
Iteration 14, loss = 0.07853949
Iteration 15, loss = 0.07383596
Iteration 16, loss = 0.06930532
Iteration 17, loss = 0.06540275
Iteration 18, loss = 0.06162943
Iteration 19, loss = 0.05837095
Iteration 20, loss = 0.05535257
Iteration 21, loss = 0.05259643
Iteration 22, loss = 0.04963190
Iteration 23, loss = 0.04767244
Iteration 24, loss = 0.04536270
Iteration 25, loss = 0.04290771
Iteration 26, loss = 0.04104809
Iteration 27, loss = 0.03941188
Iteration 28, loss = 0.03812440
Iteration 29, loss = 0.03621617
Iteration 30, loss = 0.03487377
Iteration 31, loss = 0.03334103
Iteration 32, los

MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
              beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=(100, 100), learning_rate='constant',
              learning_rate_init=0.001, max_fun=15000, max_iter=1000,
              momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
              power_t=0.5, random_state=0, shuffle=True, solver='adam',
              tol=1e-06, validation_fraction=0.1, verbose=True,
              warm_start=False)

In [11]:
# fazendo a predição usando a base de teste 
previsoes = classificador.predict(previsores_teste)

In [12]:
# observando os dados previstos com a base de teste
print(previsoes)

[1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1
 0 0 0 0 0 0 0 0 1 1 0 1 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0 1 0 0 1 0
 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 1 0
 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0
 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 1 0
 0 0 1 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 0 1 0 0 0 0 0 0 0 0 0 0 1 0
 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 0 1 1 0 0 0 0 0 0 0 0 0 0 1
 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 1 0 0 0 0 0 1 1 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0
 0 0 1 0 0 0 0 0 0 1 0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0
 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0
 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0
 0 0 0 0 0 1 0 0 0 1 0 0 

In [13]:
# observando os resultados originais para a base de teste
print(classe_teste)

[1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1
 0 0 0 0 0 0 0 0 1 1 0 1 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0 1 0 0 1 0
 0 1 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 1 0
 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0
 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 1 0
 0 0 1 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 0 1 0 0 0 0 0 0 0 0 0 0 1 0
 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 0 1 1 0 0 0 0 0 0 0 0 0 0 1
 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 1 0 0 0 0 0 1 1 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0
 0 0 1 0 0 0 0 0 0 1 0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0
 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0
 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0
 0 0 0 0 0 1 0 0 0 1 0 0 

### Analisando a capacidade de predição do algoritmo

In [15]:
# importando a biblioteca sklearn do python
from sklearn.metrics import accuracy_score, confusion_matrix

In [16]:
# obtendo a precisão do algoritmo
precisao = accuracy_score(previsoes, classe_teste)

In [17]:
# observando a precisão de acertos do algoritmo
print(precisao)

0.994


A precisão de acertos do algoritmo foi de **99.4%**.

In [18]:
# obtendo a matriz de confusão que relaciona os acertos e erros do algoritmo
matriz = confusion_matrix(previsoes, classe_teste)

In [20]:
# observando a matriz de confusão para os dados previstos pelo algoritmo
print(matriz)

[[435   2]
 [  1  62]]


Observe que 435 dados referentes a pessoas que não possuem chances de pagar o que devem foram classificados corretamente, juntamente com 62 dados referentes a pessoas que possuem chance de pagar o que devem. Entretanto, 2 dados referentes a pessoas que não tem chance de pagar foram classificados de forma errada como pagadores, assim como 1 pessoa que tem chance de pagar foi classificada de forma errada como não pagador.

### Observando o balaceamento das classes

In [21]:
# importando a biblioteca collections do python
import collections

In [22]:
# visualizando a quantidade de registros para cada uma das classes
collections.Counter(classe_teste)

Counter({1: 64, 0: 436})

### Alguma dúvida? Entre em contato comigo:

- [Me envie um e-mail](mailto:alysson.barbosa@ee.ufcg.edu.br);